paddle/fluid/framework/ir/op_fusion_pass.cc

```diff
@@ -0,0 +1,357 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <algorithm>
+#include <unordered_set>
+
+#include "paddle/fluid/framework/ir/op_fusion_pass.h"
+#include "paddle/fluid/framework/op_proto_maker.h"
+#include "paddle/fluid/framework/operator.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+const char kOpDescs[] = "op_descs";
+typedef std::vector<std::unique_ptr<OpDesc>> OpDescs;
+
+std::unique_ptr<ir::Graph> OpFusionPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  graph->Set<OpDescs>(kOpDescs, new OpDescs);
+
+  std::vector<ir::Node *> topo_order = ir::TopologySortOperations(*graph.get());
```

Comment:
The op which can be fused is not just in a topo_order.

BTW, what is the topo_order here? if there is a graph

```
a-->b
a-->c
b-->d
c-->d
```

the topo_order is `a b c d` or `a c b d` or `[a, [b, c], d]`?

Reply:
In fact, there doesn't need to get topology sort.

The topo_order is not unique, so I think the above topo_order all is right.